Validate a samplesheet automatically from a schema file

mirpedrol opened the pull request:
close #19 

This PR adds the automatic validation of a sample sheet (or any file provided through a parameter) which has a `schema` field in the parameters JSON Schema.
It also replaces the prefix "Sample" in sample sheet validation errors by "Row" to make it more general for all kinds of pipelines.


Example output of a sample sheet validation with errors:
```
ERROR ~ ERROR: Validation of 'input' file failed!

 -- Check '.nextflow.log' file for details
The following errors have been detected:

* -- Entry 1 - number: expected type: Number, found: String (string)
* -- Entry 1 - boolean: expected type: Boolean, found: String (20)
* -- Entry 1 - file: string [non_existing_file.tsv] does not match pattern ^.*\.txt$ (non_existing_file.tsv)
* -- Entry 1 - file: the file 'non_existing_file.tsv' does not exist (non_existing_file.tsv)
* -- Entry 1 - directory: the directory 'non_exsiting_folder' does not exist (non_exsiting_folder)
* -- Entry 2: Missing required value: string
* -- Entry 2: Missing required value: number
* -- Entry 2: Missing required value: boolean
* -- Entry 3 - metaBoolean: expected type: Boolean, found: String (3333)
* -- Entry 3 - metaInteger: expected type: Integer, found: String (false)

```


After discussion in this PR:

- `validateAndConvertSamplesheet()` was renamed to `fromSamplesheet()`.
- When creating a channel from a samplesheet, the validation is repeated.
- Added debug messages to allow tracking of validation time.
- fromSchema() uses the name of the parameter as argument.

mirpedrol commented:
Ah you are right, I like `Entry` more :)

nvnieuwk commented on plugins/nf-validation/src/main/nextflow/validation/SamplesheetConverter.groovy:
```diff
@@ -64,15 +60,13 @@ class SamplesheetConverter {
         JSONObject rawSchema = new JSONObject(new JSONTokener(schemaFile.text))
         SchemaLoader schemaLoader = SchemaLoader.builder()
                 .schemaJson(rawSchema)
-                .addFormatValidator("file-path", new FilePathValidator())
-                .addFormatValidator("directory-path", new DirectoryPathValidator())
```
Where are these checks now? I can't find them

mirpedrol replied:
They are in the function to validate files https://github.com/mirpedrol/nf-validation/blob/automatic-validation/plugins/nf-validation/src/main/nextflow/validation/SchemaValidator.groovy#L376
I removed it to avoid checking the same thing twice. But now that I think about, maybe it's better to validate too much than too little, I will add this back.

nvnieuwk replied:
Ah okay I missed that! It's okay if it's only there (that code is also used for the samplesheet now, right?)

mirpedrol replied:
It's also used for samplesheets yes. But could we have the case where someone wants to convert a file to a channel and doesn't validate it? It's not ideal, but I don't know if it will happen

nvnieuwk replied:
Ah yes I see what you mean hmmm. Maybe we can add some kind of check, if possible, to see if the samplesheet has been validated before converting it? So that it throws a warning to the user if they didn't validate their samplesheet?

mirpedrol replied:
I was trying to use `validateAndConvertSamplesheet()` obtaining the path from the parameters JSON schema (suggested by Phil) instead of typing the path in the pipeline. This would be ideal I think, because then it enforces you to have it specified in the schema, and it will always be validated together with parameters.

nvnieuwk replied:
Even better! That should definitely do it :)

nvnieuwk reviewed (commented):
Some small comments, but I love what you did here!

mirpedrol commented on README.md:
````diff
@@ -171,9 +171,10 @@ The path of the schema file must be relative to the root of the pipeline directo
 ```
 
-For more information about the samplesheet JSON schema refer to [samplesheet docs](docs/samplesheetDocs.md). Note that the validation performed by `validateParameters` is limited to the [JSON Schema](https://json-schema.org/) validation. Additional validation checks are performed by []`validateAndConvertSamplesheet`](https://github.com/mirpedrol/nf-validation/blob/75babb6fc293042d2c0a5acd728291bb ...
+For more information about the samplesheet JSON schema refer to [samplesheet docs](docs/samplesheetDocs.md). Note that the validation performed by `validateParameters` is limited to the [JSON Schema](https://json-schema.org/) validation. Additional validation checks are performed by []`validateAndConvertSamplesheet`](https://github.com/mirpedrol/nf-validation/blob/75babb6fc293042d2c0a5acd728291bb ...
````
Definitely 😂  thanks for the catch!

ewels commented:
```
* Missing required value: string
```
Any reason this doesn't have a line number? I think people are going to struggle to know what to do with this..

mirpedrol commented:
> Any reason this doesn't have a line number?

The validation library reports the line number together with the header name for the other errors, but for this one it only reports the name, so I can't obtain the line number from this output :(

mirpedrol commented:
> The validation library reports the line number together with the header name for the other errors, but for this one it only reports the name, so I can't obtain the line number from this output :(

Actually, I investigated a bit further, and I think I can obtain the line number. I will give it a try!

nvnieuwk commented:
> Great! Much better!
> 
> ```
> * -- Entry 1: Missing required value: string
> ```
> 
> Next question :trollface: Is it possible to get the column that is missing the required value? 🥺

It already does that :). The column is called `string` in that test 😁

ewels commented:
Ahhh, hooray! Great 😀

nvnieuwk commented:
I understand the confusion in this case 😁

ewels commented:
Ok, requests getting more nitty gritty now:

```
* -- Row 0 - directory: the directory 'non_exsiting_folder' does not exist (non_exsiting_folder)
* -- Entry 1: Missing required value: string
```
* I assume that everything is `Entry` now and that this GitHub comment is just partially updated?
* Can we make the numbering start at 1, instead of 0?

mirpedrol commented:
The comment is now updated. And entry numbers start at 1

mirpedrol commented:
Hello! 
I was thinking on the best way to deal with the sample sheet schema. I think that using the file provided in the parameters schema will add more code to the pipeline `main.nf` which is tedious. 
I also realized that `validateAndConvertSamplesheet()` doesn't use the json schema validation, so I removed the whole code for that. 
I updated the documentation accordingly, moving the definition of formats to the main README and using `params.input` instead of a file name in all examples.
I didn't change the name of the function, as I think we can still consider it validates a sample sheet.
What do you think?

ewels reviewed (commented):
Left a couple of comments, is there a reason we can't simplify usage like this?

nvnieuwk commented on README.md:
````diff
@@ -211,18 +232,19 @@ This function validates and converts a samplesheet to a ready-to-use Nextflow ch
+The function requires two different inputs: the param used by the user to provide a samplesheet and the schema used for the samplesheet. Both files need to be passed through the `file()` function as input for this function.
 
 ```nextflow
 Channel.validateAndConvertSamplesheet(
````
oooh I like this one!

mirpedrol commented on README.md:
````diff
@@ -211,18 +232,19 @@ This function validates and converts a samplesheet to a ready-to-use Nextflow ch
 ```nextflow
 Channel.validateAndConvertSamplesheet(
-   file('path/to/samplesheet', checkIfExists:true),
+   file(params.input, checkIfExists:true),
    file('path/to/schema', checkIfExists:true)
````
We can use the parameter name, but we will still need to pass the path to `nextflow_schema.json` like we do with `validateParameters()`, don't we?

ewels replied:
Can we have that as an optional argument? Seeing as it's basically always the same, should be fine to set it to a sensible default.

mirpedrol replied:
In other places we have a default if I'm not wrong, this sounds good :)

ewels commented on README.md:
```diff
@@ -211,18 +232,19 @@ This function validates and converts a samplesheet to a ready-to-use Nextflow ch
 For examples on how to process the created channel, see the [examples/](examples/) folder
 
-<!--Note that in order to fully validate the sample sheet you must always run [`validateParameters()`](https://github.com/mirpedrol/nf-validation/blob/ce409583b4582f4221cbf0d0d3917e35f4ba628d/README.md#L116) with the [`schema` field provided](https://github.com/mirpedrol/nf-validation/blob/ce409583b4582f4221cbf0d0d3917e35f4ba628d/README.md#L200).-->
+> **Note**  
+> Note that in order to fully validate the sample sheet you must always run [`validateParameters()`](#validateparameters) with the [`schema` field provided](#validate-an-input-file-provided-by-params-with-another-json-schema).
```
Ah ok, so you're saying that this channel factory doesn't validate the samplesheet? Hence you need to run `validateParameters()` to do that?

So the schema is only used in order to know the types required I guess? Maybe I need a refresher on how this all fits together..

ewels replied:
Ok, after discussion this is clear. We are using the schema to understand what the channel needs (meta fields etc), but here we are _not_ validating the samplesheet file contents against it. I think this docs note needs to mention this to explain why the `validateParameters()` needs to be called.

Alternatively, seeing as we have everything here already, it might be nice to run validation a second time - just in case parameter validation is not being run. Only reason not to do it is if it incurs a significant time penalty. If we take this route I think we can remove this note in the docs.

Would be good to have log debug messages either side of validation so that we can easily track how long it's taking.

mirpedrol commented:
According to our last discussions, I made the following changes:

- rename `validateAndConvertSamplesheet()` to `fromSamplesheet()`.
- When creating a channel from a samplesheet, the validation is repeated.
- Added debug messages to allow tracking of validation time.
- Added a default path for a samplesheet schema -> 'assets/schema_input.json'

ewels commented:
> Added a default path for a samplesheet schema -> 'assets/schema_input.json'

I thought we were taking this path from the info supplied in `nextflow_schema.json` for the supplied parameter name, no?

I'm not sure that it makes so much sense to have a default path for the sample sheet schema (parameters schema yes, but sample sheet stuff is likely to be more variable).

mirpedrol commented:
I missed that, solved now :)

ewels commented on README.md:
````diff
@@ -171,9 +191,10 @@ The path of the schema file must be relative to the root of the pipeline directo
 ```
 
-For more information about the samplesheet JSON schema refer to [samplesheet docs](docs/samplesheetDocs.md). Note that the validation performed by `validateParameters` is limited to the [JSON Schema](https://json-schema.org/) validation. Additional validation checks are performed by []`validateAndConvertSamplesheet`](https://github.com/mirpedrol/nf-validation/blob/75babb6fc293042d2c0a5acd728291bb ...
+For more information about the samplesheet JSON schema refer to [samplesheet docs](docs/samplesheetDocs.md). Note that the validation performed by `validateParameters` is limited to the [JSON Schema](https://json-schema.org/) validation. Additional validation checks are performed by [`fromSamplesheet`](#fromamplesheet).
````
```suggestion
For more information about the samplesheet JSON schema refer to [samplesheet docs](docs/samplesheetDocs.md). Note that the validation performed by `validateParameters` is limited to the [JSON Schema](https://json-schema.org/) validation. Additional validation checks are performed by [`fromSamplesheet`](#fromSamplesheet).
```

ewels commented on README.md:
```diff
@@ -205,24 +226,21 @@ This function requires an argument providing the a WorkflowMetadata object. It c
 #### Usage
 
-The function requires two different inputs: the samplesheet and the schema used for the samplesheet. Both files need to be passed through the `file()` function as input for this function.
+The function requires the name of the param used by the user to provide a samplesheet. The path to the parameters JSON schema can also be provided, defaults to '`nextflow_schema.json`').
```
```suggestion
The function requires the name of the param used by the user to provide a samplesheet. The path to the parameters JSON schema can also be provided, defaults to `nextflow_schema.json`.
```

nvnieuwk commented on README.md:
```diff
@@ -113,6 +113,26 @@ or specifying it in the configuration file
+
+| Format | Description |
+|-----------|-------------|
+| file-path | Automatically checks if the file exists and transforms the `String` type to a `Nextflow.File` type, which is usable in Nextflow processes as a `path` input |
+| directory-path | Automatically checks if the directory exists and transforms the `String` type to a `Nextflow.File` type, which is usable in Nextflow processes as a `path` input. This is currently synonymous for `file-path`. |
```
This is no longer correct (I added more types), you can check the samplesheet docs for this

nvnieuwk reviewed (approved):
One small comment otherwise it LGTM!
